File: src/Functional/Clojure/example3.clj
(ns example3)

(defprotocol ReportFactory
  (generate [this]))

(defrecord PDFReportFactory []
  ReportFactory
  (generate [_] (println "Generating PDF report")))

(defrecord HTMLReportFactory []
  ReportFactory
  (generate [_] (println "Generating HTML report")))

(defn use-factory [factory]
  (generate factory))

;; Usage
(use-factory (->PDFReportFactory))
(use-factory (->HTMLReportFactory))

File: src/Functional/Erlang/example3.erl
-module(example3).
-export([generate_report/1]).

% Abstract Factory
generate_report(pdf) -> generate_pdf_report();
generate_report(html) -> generate_html_report().

% Concrete Implementations
generate_pdf_report() -> io:format("Generating PDF report~n").
generate_html_report() -> io:format("Generating HTML report~n").

% Usage
main() ->
    generate_report(pdf),
    generate_report(html).
